# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: js/src/gc/GCInternals.h
# Commit: 2fa53cd9ac4e
# Full Hash: 2fa53cd9ac4e821763dbc4402e58b58184ce230b
# Author: Jon Coppeard <jcoppeard@mozilla.com>
# Date: 2022-03-04 04:18:28
# Regressor Bug: 1643586
# File Overlap Count: 1
# Description:
#   Bug 1643586 - Part 6: Stop using JSContext for GC helper thread tasks r=jandem
#   
#   This patch moves the debug-only GC state from JSContext to JSFreeOp. Support
#   functions for this move to FreeOp.h and most RAII classes move to GCInternals.h.
#   
# ==============================================================================

diff -r 054a08d13999 -r 2fa53cd9ac4e js/src/gc/GCInternals.h
--- a/js/src/gc/GCInternals.h	Thu Mar 03 17:49:18 2022 +0000
+++ b/js/src/gc/GCInternals.h	Thu Mar 03 17:49:18 2022 +0000
@@ -14,6 +14,7 @@
 #include "mozilla/Maybe.h"
 #include "mozilla/TimeStamp.h"
 
+#include "gc/FreeOp.h"
 #include "gc/GC.h"
 #include "vm/GeckoProfiler.h"
 #include "vm/HelperThreads.h"
@@ -239,6 +240,68 @@
   ~AutoPoisonFreedJitCode() { fop->poisonJitCode(); }
 };
 
+// Set/restore the performing GC flag for the current thread.
+class MOZ_RAII AutoSetThreadIsPerformingGC {
+  JSFreeOp* fop;
+  bool prev;
+
+ public:
+  AutoSetThreadIsPerformingGC()
+      : fop(TlsFreeOp.get()), prev(fop->isCollecting_) {
+    fop->isCollecting_ = true;
+  }
+
+  ~AutoSetThreadIsPerformingGC() { fop->isCollecting_ = prev; }
+};
+
+class MOZ_RAII AutoSetThreadGCUse {
+ protected:
+#ifndef DEBUG
+  explicit AutoSetThreadGCUse(GCUse use, JS::Zone* sweepZone = nullptr) {}
+#else
+  explicit AutoSetThreadGCUse(GCUse use, JS::Zone* sweepZone = nullptr)
+      : fop(TlsFreeOp.get()),
+        prevUse(fop->gcUse_),
+        prevZone(fop->gcSweepZone_) {
+    MOZ_ASSERT(fop->isCollecting());
+    MOZ_ASSERT_IF(sweepZone, use == GCUse::Sweeping);
+    fop->gcUse_ = use;
+    fop->gcSweepZone_ = sweepZone;
+  }
+
+  ~AutoSetThreadGCUse() {
+    fop->gcUse_ = prevUse;
+    fop->gcSweepZone_ = prevZone;
+    MOZ_ASSERT_IF(fop->gcUse() == GCUse::None, !fop->gcSweepZone());
+  }
+
+ private:
+  JSFreeOp* fop;
+  GCUse prevUse;
+  JS::Zone* prevZone;
+#endif
+};
+
+// In debug builds, update the context state to indicate that the current thread
+// is being used for GC marking.
+struct MOZ_RAII AutoSetThreadIsMarking : public AutoSetThreadGCUse {
+  explicit AutoSetThreadIsMarking() : AutoSetThreadGCUse(GCUse::Marking) {}
+};
+
+// In debug builds, update the context state to indicate that the current thread
+// is being used for GC sweeping.
+struct MOZ_RAII AutoSetThreadIsSweeping : public AutoSetThreadGCUse {
+  explicit AutoSetThreadIsSweeping(JS::Zone* zone = nullptr)
+      : AutoSetThreadGCUse(GCUse::Sweeping, zone) {}
+};
+
+// In debug builds, update the context state to indicate that the current thread
+// is being used for GC finalization.
+struct MOZ_RAII AutoSetThreadIsFinalizing : public AutoSetThreadGCUse {
+  explicit AutoSetThreadIsFinalizing()
+      : AutoSetThreadGCUse(GCUse::Finalizing) {}
+};
+
 #ifdef JSGC_HASH_TABLE_CHECKS
 void CheckHashTablesAfterMovingGC(JSRuntime* rt);
 void CheckHeapAfterGC(JSRuntime* rt);